### PR TITLE
Disable by default Portable AOT in containers for non-default gc policies

### PR DESCRIPTION
**Pull request by harryyu1994**
Issue: #10689

Signed-off-by: Harry Yu <harryyu1994@gmail.com>

### CODE REVIEW

**Comment by harryyu1994** [comment]
@dmitripivkine @mpirvu  could I get a review for this one please? I'm assuming "Gencon" is the default gc policy.

**Comment by dmitripivkine** [comment]
I guess this might be temporary work around for current situation.

However this code is a hack:
- assumes default GC policy is Gencon (correct short and medium term)
- assumes Portable AOT is generated for Gencon (can create problems on the field if there is a possibility to generate customized version of Portable AOT with different type of barriers). So customization would not work. Should customized versions of Portable AOT be not allowed?
- missing case of Concurrent Scavenger (it is under Gencon umbrella but uses different barriers).

In general GC policy is a secondary but Barriers it is using is primary criteria. Ideally Portable AOT must store "signature" of Barriers it has been generated for and check it run-time at startup to make decision about compatibility. I understand there are may be problems with early read of "signature" from pre-generated Portable AOT as well as get a current set of Barriers so early. These (and probably not only) problems should be addressed.

@mpirvu It is up for you Is there a reason for such hack. I believe we should be consistent across feature. I see a number of potential usability issues. We will need to explain customers how it works. If we can not do better for now may be we should apply some temporary restrictions.

**Comment by harryyu1994** [comment]
> Should customized versions of Portable AOT be not allowed?

With this change, customized versions of Portable AOT is still allowed. It's just that Portable AOT will not be turned on by default in containers for other gc policies. 

> missing case of Concurrent Scavenger

I guess we need the default write barrier signature to check against, instead of default gc policy.

> In general GC policy is a secondary but set of Barriers it is primary criteria. Ideally Portable AOT must store "signature" of Barriers it has been generated for and check it run-time at startup to make decision about compatibility

aotHeader->gcPolicyFlag = javaVM()->memoryManagerFunctions->j9gc_modron_getWriteBarrierType(javaVM());
JIT checks `gcPolicyFlag` to determine if AOT is compatible or not. We need a way to access shared cache before the gc component gets initialized but I'm not sure that's possible.

**Comment by mpirvu** [comment]
> @mpirvu It is up for you Is there a reason for such hack. 

@dmitripivkine I agree that the solution looks hacky, but I may have misinterpreted your intentions when you suggested to "make an early decision based on options provided about general compatibility of AOT using pre-stored set of parameters with high level of probability".

The code in this PR does the following:
1) if the user specifies -XX:+UsePortableAOT, obey that and generate portable AOT that because user knows best
2) Else, if in containers we play the probability game:
   - If the GC policy (or type of write barrier) is the default, turn on portable AOT
   - Else, do not use portable AOT

This means that an user which uses a container with portable AOT inside, but enables a different GC policy (or different write barrier) at runtime, will not use the shift value imposed by portable AOT (thus she will not take the performance penalty associated with using the forced shift value). AOT code is not going to be used anyway because the validation will fail later on.

If someone still wants to generate portable AOT inside containers with a different GC policy than the default, that is still possible, but he/she will have to specify his/her intent by specifying -XX:+UsePortableAOT.

I agree that we should look at exactly what AOT uses for validation which is` javaVM()->memoryManagerFunctions->j9gc_modron_getWriteBarrierType(javaVM())` and `javaVM()->memoryManagerFunctions->j9gc_modron_getReadBarrierType(javaVM())`. The question is whether these calls are available in `VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved)`.
In fact all complications arise from the fact that at the moment we take decisions like the compressedrefs shift value we don't have the infrastructure to validate the existing AOT header (@harryyu1994 tell me that the SCC dll is not loaded at that point in time).

File: runtime/vm/jvminit.c [hunk 1]
```diff
@@ -1814,21 +1814,23 @@ IDATA VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved) {
...
 						if (argIndexXXPortableSharedCache > argIndexXXNoPortableSharedCache) {
 							vm->extendedRuntimeFlags2 |= J9_EXTENDED_RUNTIME2_ENABLE_PORTABLE_SHARED_CACHE;
 						} else if (argIndexXXPortableSharedCache == argIndexXXNoPortableSharedCache) {
-							/* both "-XX:+PortableSharedCache" and "-XX:-PortableSharedCache" were not found thus following the default behavior for portable shared cache */
-							OMRPORT_ACCESS_FROM_J9PORT(vm->portLibrary);
-							BOOLEAN inContainer = omrsysinfo_is_running_in_container();
-							/* by default, enable portable shared cache in containers and disable otherwise */
-							if (TRUE == inContainer) {
-								vm->extendedRuntimeFlags2 |= J9_EXTENDED_RUNTIME2_ENABLE_PORTABLE_SHARED_CACHE;
+							if (J9_GC_POLICY_GENCON == ((OMR_VM *)vm->omrVM)->gcPolicy) {
```
Comment (dmitripivkine) [inline runtime/vm/jvminit.c:1822]: I am not sure the correct value is available here (so early). Would you please double test that providing non-default GC policy like `-Xgcpolicy:optthruput` changes this value?

**Pull request closed without merging**